File: base-v3/src/main/java/com/adyen/checkout/base/model/payments/request/PaymentComponentData.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 23/7/2019.
 */

package com.adyen.checkout.base.model.payments.request;

import android.os.Parcel;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.adyen.checkout.base.model.payments.Amount;
import com.adyen.checkout.core.exception.ModelSerializationException;
import com.adyen.checkout.core.model.JsonUtils;
import com.adyen.checkout.core.model.ModelObject;
import com.adyen.checkout.core.model.ModelUtils;

import org.json.JSONException;
import org.json.JSONObject;

@SuppressWarnings({"MemberName", "PMD.DataClass"})
public class PaymentComponentData<PaymentMethodDetailsT extends PaymentMethodDetails> extends ModelObject {

    @NonNull
    public static final Creator<PaymentComponentData> CREATOR = new Creator<>(PaymentComponentData.class);

    private static final String PAYMENT_METHOD = "paymentMethod";
    private static final String STORE_PAYMENT_METHOD = "storePaymentMethod";
    private static final String SHOPPER_REFERENCE = "shopperReference";
    private static final String AMOUNT = "amount";
    private static final String BILLING_ADDRESS = "billingAddress";
    private static final String DELIVERY_ADDRESS = "deliveryAddress";
    private static final String SHOPPER_NAME = "shopperName";
    private static final String TELEPHONE_NUMBER = "telephoneNumber";
    private static final String SHOPPER_EMAIL = "shopperEmail";
    private static final String DATE_OF_BIRTH = "dateOfBirth";

    @NonNull
    public static final Serializer<PaymentComponentData> SERIALIZER = new Serializer<PaymentComponentData>() {

        @NonNull
        @Override
        public JSONObject serialize(@NonNull PaymentComponentData modelObject) {
            final JSONObject jsonObject = new JSONObject();
            try {
                jsonObject.putOpt(PAYMENT_METHOD, ModelUtils.serializeOpt(modelObject.getPaymentMethod(), PaymentMethodDetails.SERIALIZER));
                jsonObject.putOpt(STORE_PAYMENT_METHOD, modelObject.isStorePaymentMethodEnable());
                jsonObject.putOpt(SHOPPER_REFERENCE, modelObject.getShopperReference());
                jsonObject.putOpt(AMOUNT, ModelUtils.serializeOpt(modelObject.getAmount(), Amount.SERIALIZER));
                jsonObject.putOpt(BILLING_ADDRESS, ModelUtils.serializeOpt(modelObject.getBillingAddress(), Address.SERIALIZER));
                jsonObject.putOpt(DELIVERY_ADDRESS, ModelUtils.serializeOpt(modelObject.getDeliveryAddress(), Address.SERIALIZER));
                jsonObject.putOpt(SHOPPER_NAME, ModelUtils.serializeOpt(modelObject.getShopperName(), ShopperName.SERIALIZER));
                jsonObject.putOpt(TELEPHONE_NUMBER, modelObject.getTelephoneNumber());
                jsonObject.putOpt(SHOPPER_EMAIL, modelObject.getShopperEmail());
                jsonObject.putOpt(DATE_OF_BIRTH, modelObject.getDateOfBirth());
            } catch (JSONException e) {
                throw new ModelSerializationException(PaymentComponentData.class, e);
            }
            return jsonObject;
        }

        @NonNull
        @Override
        public PaymentComponentData deserialize(@NonNull JSONObject jsonObject) {
            final PaymentComponentData paymentComponentData = new PaymentComponentData();
            //noinspection unchecked
            paymentComponentData.setPaymentMethod(
                    ModelUtils.deserializeOpt(jsonObject.optJSONObject(PAYMENT_METHOD), PaymentMethodDetails.SERIALIZER));
            paymentComponentData.setStorePaymentMethod(jsonObject.optBoolean(STORE_PAYMENT_METHOD));
            paymentComponentData.setShopperReference(jsonObject.optString(SHOPPER_REFERENCE));
            paymentComponentData.setAmount(ModelUtils.deserializeOpt(jsonObject.optJSONObject(AMOUNT), Amount.SERIALIZER));
            paymentComponentData.setBillingAddress(
                    ModelUtils.deserializeOpt(jsonObject.optJSONObject(BILLING_ADDRESS), Address.SERIALIZER));
            paymentComponentData.setDeliveryAddress(
                    ModelUtils.deserializeOpt(jsonObject.optJSONObject(DELIVERY_ADDRESS), Address.SERIALIZER));
            paymentComponentData.setShopperName(
                    ModelUtils.deserializeOpt(jsonObject.optJSONObject(SHOPPER_NAME), ShopperName.SERIALIZER));
            paymentComponentData.setTelephoneNumber(jsonObject.optString(TELEPHONE_NUMBER));
            paymentComponentData.setShopperEmail(jsonObject.optString(SHOPPER_EMAIL));
            paymentComponentData.setDateOfBirth(jsonObject.optString(DATE_OF_BIRTH));

            return paymentComponentData;
        }
    };

    private PaymentMethodDetailsT paymentMethod;
    private boolean storePaymentMethod;
    private String shopperReference;
    private Amount amount;
    private Address billingAddress;
    private Address deliveryAddress;
    private ShopperName shopperName;
    private String telephoneNumber;
    private String shopperEmail;
    private String dateOfBirth;

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this));
    }

    @Nullable
    public PaymentMethodDetailsT getPaymentMethod() {
        return paymentMethod;
    }

    public void setPaymentMethod(@Nullable PaymentMethodDetailsT paymentMethod) {
        this.paymentMethod = paymentMethod;
    }

    public void setStorePaymentMethod(boolean status) {
        storePaymentMethod = status;
    }

    public boolean isStorePaymentMethodEnable() {
        return storePaymentMethod;
    }

    public void setShopperReference(@Nullable String shopperReference) {
        this.shopperReference = shopperReference;
    }

    @Nullable
    public String getShopperReference() {
        return shopperReference;
    }

    @Nullable
    public Amount getAmount() {
        return amount;
    }

    public void setAmount(@Nullable Amount amount) {
        this.amount = amount;
    }

    @Nullable
    public Address getBillingAddress() {
        return billingAddress;
    }

    public void setBillingAddress(@Nullable Address billingAddress) {
        this.billingAddress = billingAddress;
    }

    @Nullable
    public Address getDeliveryAddress() {
        return deliveryAddress;
    }

    public void setDeliveryAddress(@Nullable Address deliveryAddress) {
        this.deliveryAddress = deliveryAddress;
    }

    @Nullable
    public ShopperName getShopperName() {
        return shopperName;
    }

    public void setShopperName(@Nullable ShopperName shopperName) {
        this.shopperName = shopperName;
    }

    @Nullable
    public String getTelephoneNumber() {
        return telephoneNumber;
    }

    public void setTelephoneNumber(@Nullable String telephoneNumber) {
        this.telephoneNumber = telephoneNumber;
    }

    @Nullable
    public String getShopperEmail() {
        return shopperEmail;
    }

    public void setShopperEmail(@Nullable String shopperEmail) {
        this.shopperEmail = shopperEmail;
    }

    @Nullable
    public String getDateOfBirth() {
        return dateOfBirth;
    }

    public void setDateOfBirth(@Nullable String dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }
}

File: components-core/src/main/java/com/adyen/checkout/components/core/PaymentComponentData.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 23/7/2019.
 */
package com.adyen.checkout.components.core

import com.adyen.checkout.components.core.paymentmethod.PaymentMethodDetails
import com.adyen.checkout.core.Amount
import com.adyen.checkout.core.exception.ModelSerializationException
import com.adyen.checkout.core.internal.data.model.ModelObject
import com.adyen.checkout.core.internal.data.model.ModelUtils.deserializeOpt
import com.adyen.checkout.core.internal.data.model.ModelUtils.serializeOpt
import kotlinx.parcelize.Parcelize
import org.json.JSONException
import org.json.JSONObject

/**
 * Class containing the parameters that the SDK can infer from a component's configuration and user input, especially
 * the [paymentMethod] object with the shopper input. Use [PaymentComponentData.SERIALIZER] to serialize this data to a
 * [JSONObject]. The rest of the /payments call request data should be filled in, on your server, according to your
 * needs.
 */
@Parcelize
data class PaymentComponentData<PaymentMethodDetailsT : PaymentMethodDetails>(
    var paymentMethod: PaymentMethodDetailsT?,
    var order: OrderRequest?,
    var amount: Amount?,
    var storePaymentMethod: Boolean? = null,
    var shopperReference: String? = null,
    var billingAddress: Address? = null,
    var deliveryAddress: Address? = null,
    var shopperName: ShopperName? = null,
    var telephoneNumber: String? = null,
    var shopperEmail: String? = null,
    var dateOfBirth: String? = null,
    var socialSecurityNumber: String? = null,
    var installments: Installments? = null,
    var supportNativeRedirect: Boolean? = true,
) : ModelObject() {

    companion object {
        private const val PAYMENT_METHOD = "paymentMethod"
        private const val STORE_PAYMENT_METHOD = "storePaymentMethod"
        private const val SHOPPER_REFERENCE = "shopperReference"
        private const val AMOUNT = "amount"
        private const val BILLING_ADDRESS = "billingAddress"
        private const val DELIVERY_ADDRESS = "deliveryAddress"
        private const val SHOPPER_NAME = "shopperName"
        private const val TELEPHONE_NUMBER = "telephoneNumber"
        private const val SHOPPER_EMAIL = "shopperEmail"
        private const val DATE_OF_BIRTH = "dateOfBirth"
        private const val SOCIAL_SECURITY_NUMBER = "socialSecurityNumber"
        private const val INSTALLMENTS = "installments"
        private const val ORDER = "order"
        private const val SUPPORT_NATIVE_REDIRECT = "supportNativeRedirect"

        @JvmField
        val SERIALIZER: Serializer<PaymentComponentData<*>> = object : Serializer<PaymentComponentData<*>> {
            override fun serialize(modelObject: PaymentComponentData<*>): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(PAYMENT_METHOD, serializeOpt(modelObject.paymentMethod, PaymentMethodDetails.SERIALIZER))
                        putOpt(ORDER, serializeOpt(modelObject.order, OrderRequest.SERIALIZER))
                        putOpt(AMOUNT, serializeOpt(modelObject.amount, Amount.SERIALIZER))
                        putOpt(STORE_PAYMENT_METHOD, modelObject.storePaymentMethod)
                        putOpt(SHOPPER_REFERENCE, modelObject.shopperReference)
                        putOpt(BILLING_ADDRESS, serializeOpt(modelObject.billingAddress, Address.SERIALIZER))
                        putOpt(DELIVERY_ADDRESS, serializeOpt(modelObject.deliveryAddress, Address.SERIALIZER))
                        putOpt(SHOPPER_NAME, serializeOpt(modelObject.shopperName, ShopperName.SERIALIZER))
                        putOpt(TELEPHONE_NUMBER, modelObject.telephoneNumber)
                        putOpt(SHOPPER_EMAIL, modelObject.shopperEmail)
                        putOpt(DATE_OF_BIRTH, modelObject.dateOfBirth)
                        putOpt(SOCIAL_SECURITY_NUMBER, modelObject.socialSecurityNumber)
                        putOpt(INSTALLMENTS, serializeOpt(modelObject.installments, Installments.SERIALIZER))
                        putOpt(SUPPORT_NATIVE_REDIRECT, modelObject.supportNativeRedirect)
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(PaymentComponentData::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject): PaymentComponentData<PaymentMethodDetails> {
                return PaymentComponentData(
                    paymentMethod = deserializeOpt(
                        jsonObject.optJSONObject(PAYMENT_METHOD),
                        PaymentMethodDetails.SERIALIZER
                    ),
                    order = deserializeOpt(jsonObject.optJSONObject(ORDER), OrderRequest.SERIALIZER),
                    amount = deserializeOpt(jsonObject.optJSONObject(AMOUNT), Amount.SERIALIZER),
                    storePaymentMethod = jsonObject.optBoolean(STORE_PAYMENT_METHOD),
                    shopperReference = jsonObject.optString(SHOPPER_REFERENCE),
                    billingAddress = deserializeOpt(jsonObject.optJSONObject(BILLING_ADDRESS), Address.SERIALIZER),
                    deliveryAddress = deserializeOpt(jsonObject.optJSONObject(DELIVERY_ADDRESS), Address.SERIALIZER),
                    shopperName = deserializeOpt(jsonObject.optJSONObject(SHOPPER_NAME), ShopperName.SERIALIZER),
                    telephoneNumber = jsonObject.optString(TELEPHONE_NUMBER),
                    shopperEmail = jsonObject.optString(SHOPPER_EMAIL),
                    dateOfBirth = jsonObject.optString(DATE_OF_BIRTH),
                    socialSecurityNumber = jsonObject.optString(SOCIAL_SECURITY_NUMBER),
                    installments = deserializeOpt(jsonObject.optJSONObject(INSTALLMENTS), Installments.SERIALIZER),
                    supportNativeRedirect = jsonObject.optBoolean(SUPPORT_NATIVE_REDIRECT),
                )
            }
        }
    }
}
